internal/scheduler/scheduling_algo.go

```diff
@@ -62,66 +62,81 @@ func NewFairSchedulingAlgo(
 }
 
 // Schedule assigns jobs to nodes in the same way as the old lease call.
-// It iterates over each executor in turn (using lexicographical order) and assigns the jobs using a LegacyScheduler, before moving onto the next executor
-// It maintains state of which executors it has considered already and may take multiple Schedule() calls to consider all of the executors if scheduling is slow
+// It iterates over each executor in turn (using lexicographical order) and assigns the jobs using a LegacyScheduler, before moving onto the next executor.
+// It maintains state of which executors it has considered already and may take multiple Schedule() calls to consider all executors if scheduling is slow.
 // Newly leased jobs are updated as such in the jobDb using the transaction provided and are also returned to the caller.
 func (l *FairSchedulingAlgo) Schedule(
 	ctx context.Context,
 	txn *jobdb.Txn,
 	jobDb *jobdb.JobDb,
 ) (*SchedulerResult, error) {
 	log := ctxlogrus.Extract(ctx)
-	accounting, err := l.newFairSchedulingAlgoContext(ctx, txn, jobDb)
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, l.maxSchedulingDuration)
+	defer cancel()
+
+	fsctx, err := l.newFairSchedulingAlgoContext(ctx, txn, jobDb)
 	if err != nil {
 		return nil, err
 	}
 	overallSchedulerResult := &SchedulerResult{
 		NodeIdByJobId: make(map[string]string),
 	}
-
-	timeout, cancel := context.WithTimeout(ctx, l.maxSchedulingDuration)
-	defer cancel()
-
-	allExecutorsConsidered := false
-	executorsToSchedule := accounting.getExecutorsToSchedule(l.previousScheduleClusterId)
-	for i, executor := range executorsToSchedule {
-		if timeout.Err() != nil {
-			// We've reached the scheduling time limit, exit gracefully
-			log.Infof("ending scheduling round early as we have hit the maximum scheduling duration")
-			break
+	executorGroups := l.groupExecutors(fsctx.executors)
+	if len(l.executorGroupsToSchedule) == 0 {
+		// Cycle over groups in a consistent order.
+		l.executorGroupsToSchedule = maps.Keys(executorGroups)
+		slices.Sort(l.executorGroupsToSchedule)
+	}
+	for len(l.executorGroupsToSchedule) > 0 {
+
+		select {
+		case <-ctxWithTimeout.Done():
+			// We've reached the scheduling time limit; exit gracefully.
+			log.Info("ending scheduling round early as we have hit the maximum scheduling duration")
+			return overallSchedulerResult, nil
+		default:
+		}
+		executorGroupLabel := armadaslices.Pop(&l.executorGroupsToSchedule)
+		executorGroup := executorGroups[executorGroupLabel]
+		if len(executorGroup) == 0 {
+			continue
+		}
+		for _, executor := range executorGroup {
+			if executor == nil {
+				return nil, errors.Errorf("nil executor in group %s", executorGroup)
+			}
 		}
 
-		log.Infof("scheduling on %s", executor.Id)
-		schedulerResult, sctx, err := l.scheduleOnExecutor(
-			timeout,
-			accounting,
-			txn,
-			executor,
-			jobDb,
+		// Schedule across the executors in this group.
+		// Assume pool and minimumJobSize are consistent within the group.
```

Comment:
Is this a good assumption? Definitely it historically hasn't been true all clusters in a pool have the same minimumJobSize

I think its fine, we may just need to adjust our usage

Reply:
I suppose the question is what the definition of a pool is. My intuition is that clusters in the same pool should be the same and that we should use other mechanisms (e.g., taints/tolerations) for more finer-grained control within clusters.